returnn/datasets/basic.py

```diff
@@ -960,6 +960,29 @@ def get_complete_frac(self, seq_idx):
                 num_seqs = None  # ignore
         return self.generic_complete_frac(seq_idx, num_seqs)
 
+    def get_exact_complete_frac(self, sorted_seq_idx: int) -> float:
+        """
+        Tries to calculate exactly how much of the current epoch is completed when
+        having processed seq ``sorted_seq_idx``.
+        ``sorted_seq_idx`` cannot be less than the seq index of the previously loaded seqs.
+
+        Raises a ``NotImplementedError`` if the value cannot be calculated.
+
+        This value is used to calculate ``epoch_continuous`` for any dynamic learning rate scheduling.
+
+        :param sorted_seq_idx: sorted seq idx
+        :return: continuous value in (0, 1] which represents how much of the current epoch
+            is completed after ``sorted_seq_idx``
```

Comment:
This should be extended. Specifically, you should say that we guarantee this to be monotonic (but not necessarily strictly monotonic).